Lib/test/test_asyncio/test_timeouts.py

```diff
@@ -0,0 +1,221 @@
+                    await asyncio.sleep(1)
+
+        task = asyncio.create_task(outer())
+        await task
+        self.assertFalse(task.cancelled())
+        self.assertTrue(task.done())
+
+    async def test_nested_timeouts_concurrent(self):
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(0.002):
+                with self.assertRaises(TimeoutError):
+                    async with asyncio.timeout(0.1):
+                        # Pretend we crunch some numbers.
+                        time.sleep(0.01)
+                        await asyncio.sleep(1)
+
+    async def test_nested_timeouts_loop_busy(self):
+        # After the inner timeout is an expensive operation which should
+        # be stopped by the outer timeout.
+        loop = asyncio.get_running_loop()
+        t0 = loop.time()
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(0.1):  # (1)
+                with self.assertRaises(TimeoutError):
+                    async with asyncio.timeout(0.01):  # (2)
+                        # Pretend the loop is busy for a while.
+                        time.sleep(0.1)
+                        await asyncio.sleep(1)
+                # TimeoutError was cought by (2)
+                await asyncio.sleep(10) # This sleep should be interrupted by (1)
+        t1 = loop.time()
+        self.assertTrue(t0 <= t1 <= t0 + 1)
+
+    async def test_reschedule(self):
+        loop = asyncio.get_running_loop()
+        fut = loop.create_future()
+        deadline1 = loop.time() + 10
+        deadline2 = deadline1 + 20
+
+        async def f():
+            async with asyncio.timeout_at(deadline1) as cm:
+                fut.set_result(cm)
+                await asyncio.sleep(50)
+
+        task = asyncio.create_task(f())
+        cm = await fut
+
+        self.assertEqual(cm.when(), deadline1)
+        cm.reschedule(deadline2)
+        self.assertEqual(cm.when(), deadline2)
+        cm.reschedule(None)
+        self.assertIsNone(cm.when())
+
+        task.cancel()
+
+        with self.assertRaises(asyncio.CancelledError):
+            await task
+        self.assertFalse(cm.expired())
+
+    async def test_repr_active(self):
+        async with asyncio.timeout(10) as cm:
+            self.assertRegex(repr(cm), r"<Timeout \[active\] when=\d+\.\d*>")
+
+    async def test_repr_expired(self):
+        with self.assertRaises(TimeoutError):
+            async with asyncio.timeout(0.01) as cm:
+                await asyncio.sleep(10)
+        self.assertEqual(repr(cm), "<Timeout [expired]>")
+
+    async def test_repr_finished(self):
+        async with asyncio.timeout(10) as cm:
+            await asyncio.sleep(0)
+
+        self.assertEqual(repr(cm), "<Timeout [finished]>")
+
+    async def test_repr_disabled(self):
+        async with asyncio.timeout(None) as cm:
+            self.assertEqual(repr(cm), r"<Timeout [active] when=None>")
+
```

Comment:
I'd like to add this test:
```suggestion

    async def test_nested_timeout_in_finally(self):
        with self.assertRaises(TimeoutError):
            async with asyncio.timeout(0.01):
                try:
                    await asyncio.sleep(1)
                finally:
                    with self.assertRaises(TimeoutError):
                        async with asyncio.timeout(0.01):
                            await asyncio.sleep(10)
```
It's the exact counterexample from [above](https://github.com/python/cpython/pull/31394#issuecomment-1053622308).